From: Dan Streetman <ddstreet@canonical.com>
Date: Wed, 15 Apr 2020 14:40:21 -0400
Subject: network: change UseGateway= default to UseRoutes= setting

Anyone previously using the UseRoutes=false parameter expected their
dhcp4-provided gateway route to be ignored, as well.  However, with
the introduction of the UseGateway= parameter, this is no longer true.

In order to keep backwards compatibility, this sets the UseGateway=
default value to whatever UseRoutes= has been set to.

(cherry picked from commit 589397a27759bd650b3674029cb0ef73347c913b)
---
 man/systemd.network.xml                  | 5 +++--
 src/network/networkd-network-gperf.gperf | 2 +-
 src/network/networkd-network.c           | 5 ++++-
 src/network/networkd-network.h           | 2 +-
 4 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/man/systemd.network.xml b/man/systemd.network.xml
index 120347d..5175704 100644
--- a/man/systemd.network.xml
+++ b/man/systemd.network.xml
@@ -1477,8 +1477,9 @@
         <varlistentry>
           <term><varname>UseGateway=</varname></term>
           <listitem>
-            <para>When true (the default), the gateway will be requested from the DHCP server and added to the
-            routing table with a metric of 1024, and a scope of "link".</para>
+            <para>When true, the gateway will be requested from the DHCP server and added to the routing table with a
+            metric of 1024, and a scope of "link".  When unset, the value specified with <option>UseRoutes=</option>
+            is used.</para>
           </listitem>
         </varlistentry>
         <varlistentry>
diff --git a/src/network/networkd-network-gperf.gperf b/src/network/networkd-network-gperf.gperf
index f3c7377..41fe9da 100644
--- a/src/network/networkd-network-gperf.gperf
+++ b/src/network/networkd-network-gperf.gperf
@@ -162,7 +162,7 @@ DHCPv4.UseMTU,                               config_parse_bool,
 DHCPv4.UseHostname,                          config_parse_bool,                                        0,                             offsetof(Network, dhcp_use_hostname)
 DHCPv4.UseDomains,                           config_parse_dhcp_use_domains,                            0,                             offsetof(Network, dhcp_use_domains)
 DHCPv4.UseRoutes,                            config_parse_bool,                                        0,                             offsetof(Network, dhcp_use_routes)
-DHCPv4.UseGateway,                           config_parse_bool,                                        0,                             offsetof(Network, dhcp_use_gateway)
+DHCPv4.UseGateway,                           config_parse_tristate,                                    0,                             offsetof(Network, dhcp_use_gateway)
 DHCPv4.RequestOptions,                       config_parse_dhcp_request_options,                        0,                             0
 DHCPv4.Anonymize,                            config_parse_bool,                                        0,                             offsetof(Network, dhcp_anonymize)
 DHCPv4.SendHostname,                         config_parse_bool,                                        0,                             offsetof(Network, dhcp_send_hostname)
diff --git a/src/network/networkd-network.c b/src/network/networkd-network.c
index 6b0d74c..5d2612e 100644
--- a/src/network/networkd-network.c
+++ b/src/network/networkd-network.c
@@ -265,6 +265,9 @@ int network_verify(Network *network) {
                 network->dhcp_use_mtu = false;
         }
 
+        if (network->dhcp_use_gateway < 0)
+                network->dhcp_use_gateway = network->dhcp_use_routes;
+
         if (network->dhcp_critical >= 0) {
                 if (network->keep_configuration >= 0)
                         log_warning("%s: Both KeepConfiguration= and deprecated CriticalConnection= are set. "
@@ -384,7 +387,7 @@ int network_load_one(Manager *manager, OrderedHashmap **networks, const char *fi
                 .dhcp_use_dns = true,
                 .dhcp_use_hostname = true,
                 .dhcp_use_routes = true,
-                .dhcp_use_gateway = true,
+                .dhcp_use_gateway = -1,
                 /* NOTE: this var might be overwritten by network_apply_anonymize_if_set */
                 .dhcp_send_hostname = true,
                 .dhcp_send_release = true,
diff --git a/src/network/networkd-network.h b/src/network/networkd-network.h
index a2b9cd9..fcba7e1 100644
--- a/src/network/networkd-network.h
+++ b/src/network/networkd-network.h
@@ -110,7 +110,7 @@ struct Network {
         bool dhcp_use_sip;
         bool dhcp_use_mtu;
         bool dhcp_use_routes;
-        bool dhcp_use_gateway;
+        int dhcp_use_gateway;
         bool dhcp_use_timezone;
         bool rapid_commit;
         bool dhcp_use_hostname;
